Memory game  assignment

wendellotim opened the pull request:
Hello Edwin,
I have been able to do the assignment bit which goes to the memory.js file.
However the part of index.js is still confusing me based off the instructions given. Could we start from there tomorrow is if thats okay.

Let me know. I also covered the OOP video shared and a little bit of the DOM.  I could discuss on that  as well just to demonstrate what i have learned.

Thanks

Commit:
Initial commit

Commit:
Change in the styles link

Commit:
add properties to the class

Commit:
added the card shuffle method

Commit:
add the checkifpair method

Commit:
Completed the memory.js assignment

Commit:
worked on the errors in specRunner and added the shuffle function

edwinmp commented:
@wendellotim what's the status on this? Current version is not complete. If you have any more commits, don't forget to push them.

wendellotim commented:
Hello Edwin,

So I am struggling with the code to add in the onclick function. The logic
is not playing out for me clearly. Apart from the code in the memory.js.
The js code in index.js is not adding up for me.
I have imported the class from Memory.js. The cards are shuffling. But the
other functions I am to add  the logic is failing.
I am not able to see what to do with it.

On Mon, Apr 22, 2024 at 8:53 AM Edwin P. Magezi ***@***.***>
wrote:

> @wendellotim <https://github.com/wendellotim> what's the status on this?
> Current version is not complete. If you have any more commits, don't forget
> to push them.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#issuecomment-2068544142>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSN232Z3TFJIQSCWYZKYX3Y6SQURAVCNFSM6AAAAABGCHJTTSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANRYGU2DIMJUGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
OTIM  WENDELL
0704434853
***@***.***

Commit:
Add minor refactor in the index.js

Commit:
Hide cards by default

Commit:
Add flip a card action

Commit:
add flip control on cards

edwinmp commented:
Hi @wendellotim whether tomorrow's meeting happens or not is contingent on you completing this task, or at least demonstrating reasonable progress on it. I don't see any progress since our last meeting.

wendellotim commented:
Oh yes it’s a work in progress. Before our meeting, all my solutions will
be in.

OTIM  WENDELL
0704434853
***@***.***


On Tue, 30 Apr 2024 at 16:36, Edwin P. Magezi ***@***.***>
wrote:

> Hi @wendellotim <https://github.com/wendellotim> whether tomorrow's
> meeting happens or not is contingent on you completing this task, or at
> least demonstrating reasonable progress on it. I don't see any progress
> since our last meeting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#issuecomment-2085355510>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSN2333EZGEENUABEGE5VLY76M6XAVCNFSM6AAAAABGCHJTTSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOBVGM2TKNJRGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Commit:
Compared two cards

Commit:
Add timing for the comparing

Commit:
Add event listener for the reset button

edwinmp commented on src/index.js:
```diff
@@ -1,34 +1,40 @@
+
+  document.getElementById("shuffle").addEventListener("click", () => {
+    memoryGame.shuffleCards();
+    console.log("clicked")
```
Remove the console.log. Useful for debugging but should not be committed.

edwinmp commented on src/index.js:
```diff
@@ -37,15 +43,58 @@ window.addEventListener('load', (event) => {
+      
+   **/
+
+  document.querySelector("#memory-board").innerHTML = html;
```
Add checks to account for human error in case there's no element with this id.

edwinmp replied:
You still haven't worked on this.

edwinmp commented on src/index.js:
```diff
@@ -37,15 +43,58 @@ window.addEventListener('load', (event) => {
-      console.log(`Card clicked: ${card}`);
+  document.querySelectorAll(".card").forEach((card) => {
+    card.addEventListener("click", () => {
+      if (cardsFlipped < cardLimit) {
+        card.classList.add("turned");
+        cardsFlipped++;
+
+        if (cardsFlipped === 1) {
+          firstFlippedCard = card;
+        } else if (cardsFlipped === 2) {
+          secondFlippedCard = card;
+
+          if (
+            firstFlippedCard.dataset.cardName ===
+            secondFlippedCard.dataset.cardName
+          ) {
+            setTimeout(() => {
+              firstFlippedCard.parentNode.removeChild(firstFlippedCard);
+              secondFlippedCard.parentNode.removeChild(secondFlippedCard);
+            }, 1000);
+          } else {
+            setTimeout(() => {
+              firstFlippedCard.classList.remove("turned");
+              secondFlippedCard.classList.remove("turned");
+            }, 1000);
+          }
+          cardsFlipped = 0;
+        }
```
You're not making use of the MemoryGame class, which has methods that were created to handle a lot of what you're duplicating in here, and more. That whole score board depends on the MemoryGame class.

edwinmp commented on src/index.js:
```diff
@@ -37,15 +43,58 @@ window.addEventListener('load', (event) => {
+            secondFlippedCard.dataset.cardName
+          ) {
+            setTimeout(() => {
+              firstFlippedCard.parentNode.removeChild(firstFlippedCard);
+              secondFlippedCard.parentNode.removeChild(secondFlippedCard);
```
You're removing the cards. This was not the requirement for cards that match. Just keep them face up.

edwinmp replied:
This hasn't been worked on either.

edwinmp commented on src/index.js:
```diff
@@ -37,15 +43,58 @@ window.addEventListener('load', (event) => {
+
+  document.querySelector("#memory-board").innerHTML = html;
+
+  let cardsFlipped = 0;
+  const cardLimit = 2;
+  let firstFlippedCard;
+  let secondFlippedCard;
```
You have 4 variables for something that can be done with one array that tracks the cards a user clicks. Replace this with that.

edwinmp reviewed (changes requested):
You worked on this the night before our meeting. The purpose of these meetings is not code reviews, so our meeting today is going to be very short. Since I've already introduced the concepts you're to cover next after this, there's not much to discuss. I once again recommend that you put some time aside every day to do some coding, and make sure these code reviews happen in the space between meetings, preferably with time for you to work on the feedback.

Use this task to implement linting and perhaps practice some of the debugging techniques introduced in [here](https://github.com/fourlanes-labs/wendell-otim/issues/3).

wendellotim commented:
Hello Edwin,

Thanks for the feedback, well received.

Thanks

On Wed, May 1, 2024 at 9:44 AM Edwin P. Magezi ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> You worked on this the night before our meeting. The purpose of these
> meetings is not code reviews, so our meeting today is going to be very
> short. Since I've already introduced the concepts you're to cover next
> after this, there's not much to discuss. I once again recommend that you
> put some time aside every day to do some coding, and make sure these code
> reviews happen in the space between meetings, preferably with time for you
> to work on the feedback.
>
> Use this task to implement linting and perhaps practice some of the
> debugging techniques introduced in here
> <https://github.com/fourlanes-labs/wendell-otim/issues/3>.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1585935395>
> :
>
> >  ];
>
>  const memoryGame = new MemoryGame(cards);
>
> -window.addEventListener('load', (event) => {
> -  let html = '';
> +const onload = (event) => {
> +  let html = "";
> +
> +  document.getElementById("shuffle").addEventListener("click", () => {
> +    memoryGame.shuffleCards();
> +    console.log("clicked")
>
> Remove the console.log. Useful for debugging but should not be committed.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1585935803>
> :
>
> >    // Add all the divs to the HTML
> -  document.querySelector('#memory-board').innerHTML = html;
> +  /**
> +   * Understand the game - done
> +   * Flip a card
> +   * ---Add clickevent listener on card - done
> +   * ---Add the classname turned to the class card in the car - done
> +   * Flip only two cards -- done
> +   * Compare the two cards -- done
> +   * ---If card1 and card2 are not the same flip the cards -- done
> +   * ---if Card1 and card2 are the same keep them open -- done
> +   *
> +
> +   **/
> +
> +  document.querySelector("#memory-board").innerHTML = html;
>
> Add checks to account for human error in case there's no element with this
> id.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1585936901>
> :
>
> > +      if (cardsFlipped < cardLimit) {
> +        card.classList.add("turned");
> +        cardsFlipped++;
> +
> +        if (cardsFlipped === 1) {
> +          firstFlippedCard = card;
> +        } else if (cardsFlipped === 2) {
> +          secondFlippedCard = card;
> +
> +          if (
> +            firstFlippedCard.dataset.cardName ===
> +            secondFlippedCard.dataset.cardName
> +          ) {
> +            setTimeout(() => {
> +              firstFlippedCard.parentNode.removeChild(firstFlippedCard);
> +              secondFlippedCard.parentNode.removeChild(secondFlippedCard);
> +            }, 1000);
> +          } else {
> +            setTimeout(() => {
> +              firstFlippedCard.classList.remove("turned");
> +              secondFlippedCard.classList.remove("turned");
> +            }, 1000);
> +          }
> +          cardsFlipped = 0;
> +        }
>
> You're not making use of the MemoryGame class, which has methods that were
> created to handle a lot of what you're duplicating in here, and more. That
> whole score board depends on the MemoryGame class.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1585937731>
> :
>
> > +              firstFlippedCard.parentNode.removeChild(firstFlippedCard);
> +              secondFlippedCard.parentNode.removeChild(secondFlippedCard);
>
> You're removing the cards. This was not the requirement for cards that
> match. Just keep them face up.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1585939038>
> :
>
> > +  let cardsFlipped = 0;
> +  const cardLimit = 2;
> +  let firstFlippedCard;
> +  let secondFlippedCard;
>
> You have 4 variables for something that can be done with one array that
> tracks the cards a user clicks. Replace this with that.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#pullrequestreview-2033107399>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSN235HPMADY7QSX5H5H2LZACFNLAVCNFSM6AAAAABGCHJTTSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAMZTGEYDOMZZHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
OTIM  WENDELL
0704434853
***@***.***

Commit:
Used the methods defined in the memory class

edwinmp commented:
@wendellotim have you put a proper dent in this, by which I'm asking if you're done with it. If not, perhaps we should postpone today's meeting.

Commit:
Correct indentation

wendellotim commented:
Hello Edwin,
We can have it tomorrow, same time as today.

Thanks

On Mon, May 6, 2024 at 5:01 PM Edwin P. Magezi ***@***.***>
wrote:

> @wendellotim <https://github.com/wendellotim> have you put a proper dent
> in this, by which I'm asking if you're done with it. If not, perhaps we
> should postpone today's meeting.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#issuecomment-2096090062>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSN233TNHXFREA2ONXTEJ3ZA6EMFAVCNFSM6AAAAABGCHJTTSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJWGA4TAMBWGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
OTIM  WENDELL
0704434853
***@***.***

Commit:
Add reset button

wendellotim commented:
Hello Edwin,
I added the reset button, this has an event listener, so that when we click the button, we are able to reshuffle the cards, Unfortunately Its not shuffling the cards. I added a log to show whether i am clicking, that worked but the function i am calling is not working. I wondering what I am missing

edwinmp commented:
> Hello Edwin, I added the reset button, this has an event listener, so that when we click the button, we are able to reshuffle the cards, Unfortunately Its not shuffling the cards. I added a log to show whether i am clicking, that worked but the function i am calling is not working. I wondering what I am missing

Think it through and break it down. You have to learn how to use what you know. If you can't do that, then you just know a lot of syntax. The program shouldn't be just a bunch of code to you. You should know more than just what each line is doing, but also how everything comes together. Code with intent! It's why the first thing you have to do is to start with the problem, understand what you need to do, then apply your knowledge of programming to make it happen.

So yeah, think it through and break it down - then implement each step. At this point, it doesn't matter if it's dirty - you can clean it up after.

wendellotim commented:
Hello Edwin,
This is well received.

Thanks.

OTIM  WENDELL
0704434853
***@***.***


On Tue, 7 May 2024 at 11:31, Edwin P. Magezi ***@***.***>
wrote:

> Hello Edwin, I added the reset button, this has an event listener, so that
> when we click the button, we are able to reshuffle the cards, Unfortunately
> Its not shuffling the cards. I added a log to show whether i am clicking,
> that worked but the function i am calling is not working. I wondering what
> I am missing
>
> Think it through and break it down. You have to learn how to use what you
> know. If you can't do that, then you just know a lot of syntax. The program
> shouldn't be just a bunch of code to you. You should know more than just
> what each line is doing, but also how everything comes together. Code with
> intent! It's why the first thing you have to do is to start with the
> problem, understand what you need to do, then apply your knowledge of
> programming to make it happen.
>
> So yeah, think it through and break it down - then implement each step. At
> this point, it doesn't matter if it's dirty - you can clean it up after.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#issuecomment-2097743098>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSN234VANCJUCA6VVN2STTZBCGNRAVCNFSM6AAAAABGCHJTTSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOJXG42DGMBZHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Commit:
Add shuffle button working

wendellotim commented:
Hello Edwin,

I have added the functionality now. Although it looks so bad with repetition of code. I am happy it has worked. We will optimise it I guess using functions.
Thanks

edwinmp commented:
Not "we". You give it a shot. I'd like your assurance that whatever you submit is the best you can do.

wendellotim commented:
It's best I can do but it kinda frustrates  because it's taking me long to see solutions. I have working on this since morning. hahahaha

Commit:
Add functions to simplify the code

edwinmp commented on src/index.js:
```diff
@@ -1,51 +1,102 @@
 
-window.addEventListener('load', (event) => {
-  let html = '';
+let firstFlippedCard;
+let secondFlippedCard;
+let cardsFlipped = 0;
```
As I said before, replace these with a single array.

edwinmp commented:
> Use this task to implement linting and perhaps practice some of the debugging techniques introduced in [here](https://github.com/fourlanes-labs/wendell-otim/issues/3).

You haven't done this either.

edwinmp commented:
Excellent use of functions to eliminate duplication 👍🏾

Commit:
used an array to track flippedcards in the game

Commit:
add human check on the memory board id check

wendellotim commented:
Hello Edwin,

This has been done. Thanks

On Fri, May 10, 2024 at 9:52 AM Edwin P. Magezi ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1596332181>
> :
>
> > +let firstFlippedCard;
> +let secondFlippedCard;
> +let cardsFlipped = 0;
>
> As I said before, replace these with a single array.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1596332468>
> :
>
> >    // Add all the divs to the HTML
> -  document.querySelector('#memory-board').innerHTML = html;
> +  /**
> +   * Understand the game - done
> +   * Flip a card
> +   * ---Add clickevent listener on card - done
> +   * ---Add the classname turned to the class card in the car - done
> +   * Flip only two cards -- done
> +   * Compare the two cards -- done
> +   * ---If card1 and card2 are not the same flip the cards -- done
> +   * ---if Card1 and card2 are the same keep them open -- done
> +   *
> +
> +   **/
> +
> +  document.querySelector("#memory-board").innerHTML = html;
>
> You still haven't worked on this.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1596337645>
> :
>
> > +              firstFlippedCard.parentNode.removeChild(firstFlippedCard);
> +              secondFlippedCard.parentNode.removeChild(secondFlippedCard);
>
> This hasn't been worked on either.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#pullrequestreview-2049368525>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSN234GXIAQDZULQBCBZTTZBRVB7AVCNFSM6AAAAABGCHJTTSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANBZGM3DQNJSGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
OTIM  WENDELL
0704434853
***@***.***

wendellotim commented:
I have also added a checker for the id of the memory board.

On Fri, May 10, 2024 at 9:52 AM Edwin P. Magezi ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1596332181>
> :
>
> > +let firstFlippedCard;
> +let secondFlippedCard;
> +let cardsFlipped = 0;
>
> As I said before, replace these with a single array.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1596332468>
> :
>
> >    // Add all the divs to the HTML
> -  document.querySelector('#memory-board').innerHTML = html;
> +  /**
> +   * Understand the game - done
> +   * Flip a card
> +   * ---Add clickevent listener on card - done
> +   * ---Add the classname turned to the class card in the car - done
> +   * Flip only two cards -- done
> +   * Compare the two cards -- done
> +   * ---If card1 and card2 are not the same flip the cards -- done
> +   * ---if Card1 and card2 are the same keep them open -- done
> +   *
> +
> +   **/
> +
> +  document.querySelector("#memory-board").innerHTML = html;
>
> You still haven't worked on this.
> ------------------------------
>
> In src/index.js
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#discussion_r1596337645>
> :
>
> > +              firstFlippedCard.parentNode.removeChild(firstFlippedCard);
> +              secondFlippedCard.parentNode.removeChild(secondFlippedCard);
>
> This hasn't been worked on either.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#pullrequestreview-2049368525>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSN234GXIAQDZULQBCBZTTZBRVB7AVCNFSM6AAAAABGCHJTTSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANBZGM3DQNJSGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
OTIM  WENDELL
0704434853
***@***.***

wendellotim commented:
This is the one I was super guilty of.

Thanks

On Fri, May 10, 2024 at 10:04 AM Edwin P. Magezi ***@***.***>
wrote:

> Excellent use of functions to eliminate duplication 👍🏾
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fourlanes-labs/javascript-memory-game/pull/1#issuecomment-2104021706>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJSN235UGUWKNRHCC4ALKX3ZBRWQNAVCNFSM6AAAAABGCHJTTSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMBUGAZDCNZQGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
OTIM  WENDELL
0704434853
***@***.***

edwinmp commented:
Please use GitHub instead of email to respond to the comments directly, and resolve the feedback you've worked on (there's a button for it).

edwinmp commented:
And again, request another review so that I can know you're done with everything. There's a button for it up there. Don't want to be commenting when you're still coding.
![image](https://github.com/fourlanes-labs/javascript-memory-game/assets/5672438/59c71bff-536c-45f7-9b96-f7afa2e584ec)
